Forward-port CRR replay

francoisferrand opened the pull request:
- Forward port CRR Replay implementation from 7.10 branch.
- Start all ReplayQueueProcessor instances in the same node.js process.
- Disable CRR replay from Data-Mover.

Issue: BB-234

nicolas2bert committed:
BB-117 Add instructions to deploy CRR locally

nicolas2bert committed:
BB-117 wording

Cherry-pick from 8bbe34f7a42cb2aa443ee026c5821b091ff4ed87

Issue: BB-234

miniscruff committed:
feature: BB-128 Monitor CRR replay

(cherry picked from commit 18675078beb5f9ec92f3051c54401b49a7758dc9)

Issue: BB-234

francoisferrand committed:
Implement CRR Replay

Cherry-picked from 0053b2816fa718f5cf54a204402d8582f417b3dc

Issue: BB-234

francoisferrand committed:
Fix lint issues

Issue: BB-234

francoisferrand committed:
Fix failing test due to new replayCount field

Issue: BB-234

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-234 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.10`


Please check the `Fix Version/s` of BB-234, or the target
branch of this pull request.

francoisferrand committed:
Fix failing replication test

Due to updated QueueProcessor api.

Issue: BB-234

francoisferrand commented on extensions/replication/tasks/UpdateReplicationStatus.js:
```diff
@@ -277,7 +371,7 @@ class UpdateReplicationStatus extends BackbeatTask {
             const updatedSourceEntry = this._getUpdatedSourceEntry(params, log);
             if (!updatedSourceEntry) {
-                return done();
+                return process.nextTick(done);
```
Original code had this 'nextTick()', but not sure it is still relevant in the updated code

@nicolas2bert Can you provide more context/feedback?

nicolas2bert replied:
I will keep it to keep the callback function asynchronous in both cases.

francoisferrand replied:
thanks

francoisferrand committed:
Run all ReplayProcessors in the same process

Change the deployment model for replay processors, from 1 process per
target (S3C/7.x) to 1 process for all targets (Zenko/8.x), like it is
done for the queue processor.

Since the task.js entrypoint is the same, deduplicate the code to keep
only the one from QueueProcessor, by taking an optional param on command
line to specify the "replayTopic" to process: if not specified, it will
simply process the replication topic (as before), otherwise it will be a
replay processor.

Issue: BB-234

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-234 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.11`


Please check the `Fix Version/s` of BB-234, or the target
branch of this pull request.

francoisferrand committed:
Use separate zookeeper path for replay processors

The state is managed separate, to avoid any race condition; however the
QueueProcessor and ReplayProcessor's watch the same redis channel
(associated with the topic) : so they will be paused/resumed together
when backbeat API is used.

Issue: BB-234

francoisferrand committed:
Consume DataMover topic only in queue processor

Replay processors should not attempt to read from data mover topic.

Issue: BB-234

nicolas2bert commented on package.json:
```diff
@@ -15,6 +15,7 @@
     "oplog_populator": "node extensions/oplogPopulator/OplogPopulatorTask.js",
     "mongo_queue_processor": "node extensions/mongoProcessor/mongoProcessorTask.js",
     "garbage_collector": "node extensions/gc/service.js",
+    "replay_processor": "node extensions/replication/replayProcessor/task.js",
```
Not sure it is needed anymore.

nicolas2bert commented on extensions/replication/queueProcessor/QueueProcessor.js:
```diff
@@ -628,6 +637,10 @@ class QueueProcessor extends EventEmitter {
 
     _getZkSiteNode() {
+        if (this.isReplayTopic) {
+            return `${zookeeperNamespace}${zkReplayStatePath}/${this.topic}/${this.site}`;
```
Today, `${zookeeperNamespace}${zkReplayStatePath}/${this.topic}/${this.site}` and `${zookeeperNamespace}${zkStatePath}/${this.site}` zk paths will share the same state. 
Pausing and Resuming a site will pause/resume both the site replication and the site replay work. 
I think it is the expected behavior, just wanted to double check with you :)

francoisferrand replied:
We do keep separate states in zookeeper on purpose, to avoid race conditions (eg when updating the state);
but every instance will react to the same "commands" (eg pause/resume a site), so their states should be the same indeed.

nicolas2bert reviewed (commented):
Also, are we planning on having some CRR Replay monitoring? @miniscruff worked on it for S3C if you want to have a look :) : https://github.com/scality/backbeat/pull/2086

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-234 contains:


* `8.4.12`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.14`


Please check the `Fix Version/s` of BB-234, or the target
branch of this pull request.

francoisferrand committed:
Remove unneeded replay_processor task

Issue: BB-234

francoisferrand commented:
> Also, are we planning on having some CRR Replay monitoring? @miniscruff worked on it for S3C if you want to have a look :) : https://github.com/scality/backbeat/pull/2086 

Added this code, though we will add metrics retrieval and dashboard separately: there is no replication monitoring at the moment in zenko, it will be addressed in separate epic.

francoisferrand committed:
Allow injecting replication errors

Issue: BB-234

welansari reviewed (approved):
LGTM, tested locally

Commit:
bump backbeat 8.4.15

francoisferrand commented:
/approve

Kerkesni commented:
ping

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

erwan-b commented:
ping

francoisferrand commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-234 contains:


* `8.4.15`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.19`

* `8.4.15`


Please check the `Fix Version/s` of BB-234, or the target
branch of this pull request.

tcarmet commented:
don't mind me just testing stuff.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-234.

Goodbye francoisferrand.
